node/src/pruner.rs

```diff
@@ -0,0 +1,272 @@
+
+// pruning window is 1 hour behind syncing window
+const PRUNING_WINDOW: Duration = SYNCING_WINDOW.saturating_add(Duration::from_secs(60 * 60));
+const PRUNING_INTERVAL: Duration = Duration::from_secs(60);
+
+type Result<T, E = PrunerError> = std::result::Result<T, E>;
+
+/// Representation of all the errors that can occur when interacting with the [`Pruner`].
+#[derive(Debug, thiserror::Error)]
+pub enum PrunerError {
+    /// An error propagated from the [`P2p`] module.
+    #[error(transparent)]
+    P2p(#[from] P2pError),
+
+    /// An error propagated from the [`Store`] module.
+    #[error(transparent)]
+    Store(#[from] StoreError),
+
+    #[error(transparent)]
+    Blockstore(#[from] blockstore::Error),
+}
+
+pub struct Pruner {
+    cancellation_token: CancellationToken,
+}
+
+pub struct PrunerArgs<S, B>
+where
+    S: Store,
+    B: Blockstore,
+{
+    /// Headers storage.
+    pub store: Arc<S>,
+    /// Block storage.
+    pub blockstore: Arc<B>,
+    /// Event publisher.
+    pub event_pub: EventPublisher,
+}
+
+impl Pruner {
+    pub fn start<S, B>(args: PrunerArgs<S, B>) -> Self
+    where
+        S: Store + 'static,
+        B: Blockstore + 'static,
+    {
+        let cancellation_token = CancellationToken::new();
+        let event_pub = args.event_pub.clone();
+
+        let mut worker = Worker::new(args, cancellation_token.child_token());
+
+        spawn(async move {
+            if let Err(e) = worker.run().await {
+                error!("Pruner stopped because of a fatal error: {e}");
+
+                event_pub.send(NodeEvent::FatalPrunerError {
+                    error: e.to_string(),
+                });
+            }
+        });
+
+        Pruner { cancellation_token }
+    }
+
+    pub fn stop(&self) {
+        self.cancellation_token.cancel();
+    }
+}
+
+impl Drop for Pruner {
+    fn drop(&mut self) {
+        self.cancellation_token.cancel();
+    }
+}
+
+struct Worker<S, B>
+where
+    S: Store + 'static,
+    B: Blockstore + 'static,
+{
+    cancellation_token: CancellationToken,
+    event_pub: EventPublisher,
+    store: Arc<S>,
+    blockstore: Arc<B>,
+}
+
+impl<S, B> Worker<S, B>
+where
+    S: Store,
+    B: Blockstore,
+{
+    fn new(args: PrunerArgs<S, B>, cancellation_token: CancellationToken) -> Self {
+        Worker {
+            cancellation_token,
+            event_pub: args.event_pub,
+            store: args.store,
+            blockstore: args.blockstore,
+        }
+    }
+
+    async fn run(&mut self) -> Result<()> {
+        let mut last_reported = None;
+        let mut last_removed = None;
+        loop {
+            let pruning_window_end = Time::now().checked_sub(PRUNING_WINDOW).unwrap_or_else(|| {
+                warn!("underflow when computing pruning window start, defaulting to unix epoch");
+                Time::unix_epoch()
+            });
+
+            while let Some(header) = self.get_tail_header_to_prune(&pruning_window_end).await? {
```

Comment:
After `retain` is implemented. It would make sense to do a binary search to find up to which height we should remove.

Reply:
Per conversation during daily, we'll go with one-by-one removal.